components/places/src/db/db.rs

```diff
@@ -9,22 +9,123 @@
 use super::interrupt::{InterruptScope, PlacesInterruptHandle};
 use super::schema;
 use crate::error::*;
-use rusqlite::Connection;
+use lazy_static::lazy_static;
+use rusqlite::{Connection, OpenFlags};
 use sql_support::ConnExt;
+use std::cell::Cell;
 use std::ops::Deref;
-use std::path::Path;
+use std::path::{Path, PathBuf};
+use std::sync::Mutex;
 
 use std::sync::{atomic::AtomicUsize, Arc};
 
 pub const MAX_VARIABLE_NUMBER: usize = 999;
 
+#[derive(Debug, PartialEq)]
+pub enum ConnectionType {
+    ReadOnly,
+    ReadWrite,
+    Sync,
+}
+
+impl ConnectionType {
+    fn rusqlite_flags(&self) -> OpenFlags {
+        let common_flags = OpenFlags::SQLITE_OPEN_NO_MUTEX | OpenFlags::SQLITE_OPEN_URI;
+        match self {
+            ConnectionType::ReadOnly => common_flags | OpenFlags::SQLITE_OPEN_READ_ONLY,
+            ConnectionType::ReadWrite => {
+                common_flags | OpenFlags::SQLITE_OPEN_CREATE | OpenFlags::SQLITE_OPEN_READ_WRITE
+            }
+            ConnectionType::Sync => common_flags | OpenFlags::SQLITE_OPEN_READ_WRITE,
+        }
+    }
+}
+
+// XXX - move this struct to ./api.rs?
+/// The entry-point to the places API. This object gives access to database
+/// connections and other helpers. It enforces that only 1 write connection
+/// can exist to the database at once.
+pub struct PlacesAPI {
+    db_name: PathBuf,
+    encryption_key: Option<String>,
+    write_connection: Mutex<Cell<Option<PlacesDb>>>,
+    // XXX - TODO - the "sync" connection.
+}
+
+impl PlacesAPI {
+    /// Create a new PlacesAPI backed by a file on disk.
+    pub fn new(db_name: impl AsRef<Path>, encryption_key: Option<&str>) -> Result<Self> {
+        // We always create a new read-write connection for an initial open so
+        // we can create the schema and/or do version upgrades.
+        // XXX - pathbuf usage here is wrong.
```

Comment:
This should probably be `pub fn new(db_name: impl Into<PathBuf>, ...)`
and then
```rust
let db_name = db_name.into();
let connection = PlacesDb::open(&db_name, ...);
...
Ok(PlacesAPI { db_name, ... })
```